tools/gcc-bridge/compiler/src/main/java/org/renjin/gcc/codegen/constructors/Constructors.java

```diff
@@ -0,0 +1,287 @@
+        return null;
+      }
+      GimpleStringConstant stringConstant = (GimpleStringConstant) addressOf.getValue();
+      if (!isNullTerminated(stringConstant)) {
+        return null;
+      }
+
+      stringBuilder.append(stringConstant.getValue());
+    }
+
+    if (stringBuilder.length() < 1000) {
+
+      // Store the concatenated string in the Class' string pool
+
+      mv.aconst(stringBuilder.toString());
+      mv.invokestatic(Type.getType(BytePtr.class), "stringArray",
+          Type.getMethodDescriptor(Type.getType(Ptr.class), Type.getType(String.class)));
+
+    } else {
+
+      // Store in an external resource
+
+      byte[] bytes = stringBuilder.toString().getBytes(Charsets.US_ASCII);
+      String resourceName = Hashing.md5().hashBytes(bytes).toString();
+
+      try {
+        resourceWriter.writeResource(resourceName, bytes);
+      } catch (IOException e) {
+        throw new InternalCompilerException("Failed to write resource", e);
+      }
+
+      mv.aconst(mv.getOwnerClass());
+      mv.aconst(resourceName);
+      mv.invokestatic(Type.getType(BytePtr.class), "stringArrayFromResource",
+          Type.getMethodDescriptor(Type.getType(Ptr.class), Type.getType(Class.class), Type.getType(String.class)));
+    }
+
+    // Save the array to a local variable
+
+    LocalVar var = mv.getLocalVarAllocator().reserve(Type.getType(Ptr.class));
+
+    mv.store(var.getIndex(), Type.getType(Ptr.class));
+
+    return new VArrayExpr(arrayType, new VPtrExpr(var));
+  };
+
+  static ConstructorInterface charArray = (mv, resourceWriter, expr) -> {
+
+    if (!isArrayWithType(expr.getType(), c -> isIntegerWithPrecision(c, 16))) {
+      return null;
+    }
+
+    StringBuilder s = new StringBuilder();
+    for (GimpleConstructor.Element element : expr.getElements()) {
+      GimpleIntegerConstant constant = (GimpleIntegerConstant) element.getValue();
+      s.appendCodePoint(constant.getNumberValue().intValue());
+    }
+
+    JExpr array = new JExpr() {
+
+      @Nonnull
+      @Override
+      public Type getType() {
+        return Type.getType(Ptr.class);
+      }
+
+      @Override
+      public void load(@Nonnull MethodGenerator mv) {
+        mv.aconst(s.toString());
+        mv.invokestatic(Type.getType(CharPtr.class), "int16Array",
+            Type.getMethodDescriptor(Type.getType(Ptr.class),
+                Type.getType(String.class)));
+
+      }
+    };
+
+    return new VArrayExpr((GimpleArrayType) expr.getType(), new VPtrExpr(array));
+  };
+
+  /* Convenience functions */
+
+  private static ByteBuffer constantArrayToBuffer(
+      GimpleConstructor value,
+      Predicate<GimpleType> isElementType,
+      BufferConverter converter
+  ) {
+    boolean isArrayType = isArrayWithType(value.getType(), isElementType);
+    if (!isArrayType) {
+      return null;
+    }
+    List<GimpleConstructor.Element> elements = value.getElements();
+    if (elements.size() < 4000) { // only stream large arrays
```

Comment:
This limit might be high. The bytecode for initializing an array looks (something) like this:

```
NEWARRAY int
DUP              <- duplicate the array reference on the stack (1 byte)
ICONST_0   <- array index (1-3 bytes??)
ICONST ...  <- array value (1-3 bytes)
IASTORE    <- store value in array on stack
DUP          
ICONST_1 
BIPUSH ...
IASTORE
DUP
ICONST <n>
BIPUSH ...
ASTORE  <varindex>
```
so storing a single 32-bit integer element can require up to 6 bytes of instructions _and_ a constant pool entry for the integer if doesn't fit into the SIPUSH range. The limit for strings was higher because there's facility for storing an entire string in the constant pool.

Reply:
No problems.  What limit would you recommend?  Just use 1000?

Reply:
I'll cap it to 128, which limits the damage to ~1KiB

Reply:
I'm seeing that the output for doubles is something like this:

```
    DUP (1B)
    BIPUSH 98 (2B)
    LDC 98.0 (9B)
    DASTORE (1B)
```
... for the first 128, and like this thereafter:
```
    DUP (1B)
    SIPUSH 8199 (3B)
    LDC 4099.0 (9B)
    DASTORE (1B)
```

So 14 bytes per entry.  Which is strange because it takes an array of size 8209 doubles to elicit the `Method code too large!` issue, which calculates to:
```
(8209-128) * 14 BIPUSH
+ 128 * 13 SIPUSH
- (5*2 + 2*8) savings from constants
= 114,772 byte
```
... well in excess of the 65536 class code size limit.  Any idea why that might be?  Is there some pre-optimisation happening?

I'm also concerned about the use of constant pool for arrays.  For the string implementation I notice that it doesn't check the free space of the class constant pool, so theoretically (though unlikely) it would be possible for enough <1000 size arrays to exceed the 65535 byte limit of the constant pool.

Reply:
Sorry, I'm learning the JVM as I go 🙂  I see now that the run-time constant pool has it's own separate limitation. This doesn't change my concern that a bug lives in the constant pooling mechanism.

Reply:
The ASM library does offer a very thin layer of abstraction over some of the bytecodes. So `LDC <double>` is written as `ldc_w indexbyte1 indexbyte2` where the indexbytes refer to a constant pool entry:
https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-6.html#jvms-6.5.ldc_w

So at most, an array element will require 8 bytes of bytecode, which clocks in at 64000 for 8000 elements. Then there are savings as you said when BIPUSH or ICONST_0...5 can be used for the index, or for DCONST_0...1 can be used for doubles. In principle, we don't need initialize zero-valued array elements but I'm not sure if GCC-Bridge currently takes that into account.

You can inspect the generated classfiles using the `javap` commandline tool:

```
javap -c -p com/acme/MyClass.class
```
I think the `-v` option will dump the constant pool as well.

Reply:
Thanks @akbertram 🙂   I was reading the spec last night and came to the same understanding, and I've updated the comments in `arrays.c` to match.  So far I've been using the debug logging in the test results, but that `javap` looks really useful.

I'm not sure if you missed my question, but I was wondering if there's a way I can generate a local gcc-build maven plugin?  Such that I can build my c library in the same manner as `gcc-bridge-example` with this branch code?  Otherwise I'll try generate my own script with the generated jars.